models/tests/test_document.py

```diff
@@ -0,0 +1,168 @@
+from unittest import TestCase
+
+from models.base.document import IndexedDocument, Document
+from models.base.field import Field, ReferenceDocumentsField, ReferenceSet
+from models.base.meta_document import MetaDocument
+
+
+class MetaDocumentTest(TestCase):
+    def test_multiple_primary_key(self):
+        with self.assertRaises(MetaDocument.MultiplePrimaryKeyError):
+            class DemoIndexedDocument(IndexedDocument):
+                id = Field(primary_key=True)
+                name = Field(primary_key=True)
+
+
+class BasicDocumentTest(TestCase):
+    class DemoIndexedDocument(IndexedDocument):
+        name = Field(primary_key=True)
+        password = Field()
+
+    def test_create_document(self):
+        doc = self.DemoIndexedDocument(name='test', password='test')
+        self.assertEqual(doc.name, 'test')
+        self.assertEqual(doc.password, 'test')
+
+    def test_compulsory_primary_key(self):
+        with self.assertRaises(Document.PrimaryKeyNotSetError):
+            self.DemoIndexedDocument(password='test')
+
+
+class DocumentPersistenceTest(TestCase):
+    class DemoIndexedDocument(IndexedDocument):
+        id = Field(primary_key=True)
+        name = Field()
+
+    def setUp(self) -> None:
+        self.DemoIndexedDocument(id=1, name='a')
+        self.DemoIndexedDocument(id=2, name='b')
+        self.DemoIndexedDocument(id=3, name='c')
+
+    def test_persistence(self):
+        self.DemoIndexedDocument.reload()
+        self.assertEqual(len(self.DemoIndexedDocument.all()), 3)
+
+    def test_find(self):
+        document = self.DemoIndexedDocument.find(1)
+        self.assertEqual(document.name, 'a')
+        document = self.DemoIndexedDocument.find(10)
+        self.assertIsNone(document)
+
+    def test_delete(self):
+        self.DemoIndexedDocument(id=4, name='d')
+        document = self.DemoIndexedDocument.find(4)
+        self.assertIsNotNone(document)
+        document.delete()
+        self.DemoIndexedDocument.reload()
+        document = self.DemoIndexedDocument.find(4)
+        self.assertIsNone(document)
+
+    def tearDown(self) -> None:
+        self.DemoIndexedDocument.delete_all()
+
+
+class DocumentReferenceTest(TestCase):
+    class DemoDocument(Document):
+        name = Field()
+
+    class DemoNestedDocument(IndexedDocument):
+        name = Field(primary_key=True)
+        children = ReferenceDocumentsField()
+
+    def setUp(self) -> None:
+        children = [
+            self.DemoDocument(name='a'),
+            self.DemoDocument(name='b'),
+            self.DemoDocument(name='c')
+        ]
+        self.document = self.DemoNestedDocument(name='test', children=children)
+
+    def test_retrieval(self):
+        self.assertEqual(3, len(self.document.children))
+
+    def test_add(self):
+        self.document.children.add(self.DemoDocument(name='d'))
+        self.assertEqual(4, len(self.document.children))
+
+    def test_add_incorrect_type(self):
```

Comment:
Is this testing that if you try to add children that are already present then it raises the MultipleTypeError? 
If not, what does an incorrect type mean exactly?

Reply:
Incorrect type means that you are adding documents of different types to the same `ReferenceDocumentsField`
In the test, `self.document` was first created with children of type `DemoDocument`. 
In this test, it was attempted to add another child of type `DemoNestedDocument` to the same field, hence the error.